test/src/main/java/org/apache/accumulo/test/fate/accumulo/FateStoreIT.java

```diff
@@ -217,10 +228,144 @@ protected void testDeferredOverflow(FateStore<TestEnv> store, ServerContext sctx
+    FateKey fateKey = FateKey.forSplit(ke);
+    FateId fateId1 = store.create(fateKey);
+    FateId fateId2 = store.create(fateKey);
+    assertEquals(fateId1, fateId2);
+
+    FateTxStore<TestEnv> txStore = store.reserve(fateId1);
+    try {
+      assertTrue(txStore.timeCreated() > 0);
+      assertEquals(TStatus.NEW, txStore.getStatus());
+      assertEquals(fateKey, txStore.getKey().orElseThrow());
+      assertEquals(1, store.list().count());
+    } finally {
+      txStore.delete();
+      txStore.unreserve(0, TimeUnit.SECONDS);
+    }
+  }
+
+  @Test
+  public void testCreateWithKeyInProgress() throws Exception {
+    executeTest(this::testCreateWithKeyInProgress);
+  }
+
+  protected void testCreateWithKeyInProgress(FateStore<TestEnv> store, ServerContext sctx) {
+    KeyExtent ke = new KeyExtent(TableId.of("tableId"), new Text("zzz"), new Text("aaa"));
+
+    FateKey fateKey = FateKey.forSplit(ke);
+    FateId fateId = store.create(fateKey);
+
+    FateTxStore<TestEnv> txStore = store.reserve(fateId);
+    try {
+      assertTrue(txStore.timeCreated() > 0);
+      txStore.setStatus(TStatus.IN_PROGRESS);
+
+      // We have an existing transaction with the same key in progress
+      // so should not be allowed
+      assertThrows(IllegalStateException.class, () -> store.create(fateKey));
+    } finally {
+      txStore.delete();
+      txStore.unreserve(0, TimeUnit.SECONDS);
+    }
+
+    try {
+      // After deletion, make sure we can create again with the same key
+      fateId = store.create(fateKey);
+      txStore = store.reserve(fateId);
+      assertTrue(txStore.timeCreated() > 0);
+      assertEquals(TStatus.NEW, txStore.getStatus());
+    } finally {
+      txStore.delete();
+      txStore.unreserve(0, TimeUnit.SECONDS);
+    }
+
+  }
+
+  @Test
+  public void testCreateWithKeyCollision() throws Exception {
+    // Replace the default hasing algorithm with one that always returns the same tid so
+    // we can check duplicate detection with different keys
+    executeTest(this::testCreateWithKeyCollision, AbstractFateStore.DEFAULT_MAX_DEFERRED,
+        (instanceType, fateKey) -> FateId.from(instanceType, 1000));
+  }
+
+  protected void testCreateWithKeyCollision(FateStore<TestEnv> store, ServerContext sctx) {
+    KeyExtent ke1 = new KeyExtent(TableId.of("tableId1"), new Text("zzz"), new Text("aaa"));
+    KeyExtent ke2 = new KeyExtent(TableId.of("tableId2"), new Text("ddd"), new Text("bbb"));
+
+    FateKey fateKey1 = FateKey.forSplit(ke1);
+    FateKey fateKey2 = FateKey.forSplit(ke2);
+    FateId fateId1 = store.create(fateKey1);
+
+    FateTxStore<TestEnv> txStore = store.reserve(fateId1);
+    try {
+      try {
+        store.create(fateKey2);
+        fail("Expected IllegalStateException due to hashing collision");
+      } catch (Exception e) {
+        assertInstanceOf(IllegalStateException.class, e);
+        assertEquals("Collision detected for tid 1000", e.getMessage());
+      }
```

Comment:
Thanks for the suggestion, I was trying to use assertThrows() but I wanted to check the message and didn't see a way to do that with assertThrows(). I didn't realize that method also returns the exception so that makes it easy to check.

Reply:
>  I didn't realize that method also returns the exception so that makes it easy to check.

Yeah I didn't know that for a while either when I was using it.  It is a really nice design by the junit team.